fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -1848,6 +1850,48 @@ Future<Optional<std::string>> AdvanceVersionImpl::commit(ReadYourWritesTransacti
 	return Optional<std::string>();
 }
 
+ACTOR static Future<RangeResult> getVersionEpochActor(ReadYourWritesTransaction* ryw, KeyRangeRef kr) {
+	ryw->getTransaction().setOption(FDBTransactionOptions::LOCK_AWARE);
+	ryw->getTransaction().setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+	Optional<Value> val = wait(ryw->getTransaction().get(versionEpochKey));
+	RangeResult result;
+	if (val.present()) {
+		int64_t versionEpoch = BinaryReader::fromStringRef<Version>(val.get(), Unversioned());
+		ValueRef version(result.arena(), boost::lexical_cast<std::string>(versionEpoch));
+		result.push_back_deep(result.arena(), KeyValueRef(kr.begin, version));
+	}
+	return result;
+}
+
+VersionEpochImpl::VersionEpochImpl(KeyRangeRef kr) : SpecialKeyRangeRWImpl(kr) {}
+
+Future<RangeResult> VersionEpochImpl::getRange(ReadYourWritesTransaction* ryw, KeyRangeRef kr) const {
+	ASSERT(kr == getKeyRange());
+	return getVersionEpochActor(ryw, kr);
+}
+
+ACTOR static Future<Optional<std::string>> versionEpochCommitActor(ReadYourWritesTransaction* ryw, Version v) {
+	ryw->getTransaction().setOption(FDBTransactionOptions::LOCK_AWARE);
+	ryw->getTransaction().setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+	ryw->getTransaction().set(versionEpochKey, BinaryWriter::toValue(v, Unversioned()));
+	return Optional<std::string>();
+}
+
+Future<Optional<std::string>> VersionEpochImpl::commit(ReadYourWritesTransaction* ryw) {
+	auto versionEpoch =
+	    ryw->getSpecialKeySpaceWriteMap()[SpecialKeySpace::getManagementApiCommandPrefix("versionepoch")].second;
+	if (versionEpoch.present()) {
+		try {
+			int64_t v = boost::lexical_cast<int64_t>(versionEpoch.get().toString());
+			return versionEpochCommitActor(ryw, v);
+		} catch (boost::bad_lexical_cast& e) {
+			return Optional<std::string>(ManagementAPIError::toJsonString(
+			    false, "versionepoch", "Invalid epoch (int64_t) argument: " + versionEpoch.get().toString()));
+		}
+	}
```

Comment:
Does this mean clearing the version epoch special key has no effect? I don't have a strong preference one way or the other for this, it could be disallowed or it could revert to the knob based default epoch. I think if we choose to disable it, we should override the `clear` functions here and throw an error if the key gets cleared.

Reply:
Assuming a typical approach is going from a cluster that does not have a version epoch set, to the cluster using the default version epoch (of Jan 1, 2010). This instantly causes a ~3.7 * 10^14 version jump. If the operator decides they didn't mean to do that, trying to reset things will not really be tenable. So I don't think a clear should be allowed on the key.

A clear meaning the version epoch is reset to the default value could be done, but it's a little confusing with the semantics of the version epoch being an optional. When I think of clearing the key, I think of the optional being reset, thus causing there to be no version epoch, not being reset to the default version epoch. So I think removing clear functionality makes sense here.

Reply:
How does one signal to the cluster that it should use the default version epoch? I was under the impression that this happened if the key was unset, but maybe that's not accurate.

Reply:
If the key in unset, the old way of calculating versions is used, and this change should have no effect. Currently, the way to switch to the default version epoch is to run `versionepoch add 0`, which sets the key and adds 0 to the current version epoch (which, when unset, will default to Jan 1, 2010). Running `versionepoch set 1262307600` is another way, but that requires knowing the default value.

Reply:
Ok, got it. In that model, clearing the key could turn off time-synchronization without making any changes to the current version, which seems like it could be a sensible operation.

Reply:
I see, yeah that sounds reasonable.